sdks/python/apache_beam/runners/dataflow/dataflow_metrics.py

```diff
@@ -21,13 +21,88 @@
 service.
 """
 
+from collections import defaultdict
+
+from apache_beam.metrics.execution import MetricKey
+from apache_beam.metrics.execution import MetricResult
 from apache_beam.metrics.metric import MetricResults
+from apache_beam.metrics.metricbase import MetricName
 
 
 # TODO(pabloem)(JIRA-1381) Implement this once metrics are queriable from
 # dataflow service
 class DataflowMetrics(MetricResults):
+  """Implementation of MetricResults class for the Dataflow runner."""
+
+  def __init__(self, dataflow_client=None, job_result=None):
+    """Initialize the Dataflow metrics object.
+
+    Args:
+      dataflow_client: apiclient.DataflowApplicationClient to interact with the
+        dataflow service.
+      job_result: DataflowPipelineResult with the state and id information of
+        the job
+    """
+    super(DataflowMetrics, self).__init__()
+    self._dataflow_client = dataflow_client
+    self.job_result = job_result
+    self._queried_after_termination = False
+    self._cached_metrics = None
+
+  def _populate_metric_results(self, response):
+    """Take a list of metrics, and convert it to a list of MetricResult."""
+    user_metrics = [metric
+                    for metric in response.metrics
+                    if metric.name.origin == 'user']
+
+    # Get the tentative/committed versions of every metric together.
+    metrics_by_name = defaultdict(lambda: {})
+    for metric in user_metrics:
+      tentative = [prop
+                   for prop in metric.name.context.additionalProperties
+                   if prop.key == 'tentative' and prop.value == 'true']
+      key = 'tentative' if tentative else 'committed'
+      metrics_by_name[metric.name.name][key] = metric
+
+    # Now we create the MetricResult elements.
+    result = []
+    for name, metric in metrics_by_name.iteritems():
+      if (name.endswith('(DIST)') or
+          name.endswith('[MIN]') or
+          name.endswith('[MAX]') or
+          name.endswith('[MEAN]') or
+          name.endswith('[COUNT]')):
+        # Distributions are not yet fully supported in this runner
```

Comment:
There's a TODO on line 106. Also, distributions will be shown in the UI, but they won't be returned by the `query` method.  I've added a warning.